src/service/angular-sails.js

```diff
@@ -1,74 +1,168 @@
-            },
-            angularify = function (cb, data) {
-                $timeout(function () {
-                    cb(data);
-                });
-            },
-            promisify = function (methodName) {
-                socket['legacy_' + methodName] = socket[methodName];
-                socket[methodName] = function (url, data, cb) {
-                    var deferred = defer();
-                    if (cb === undefined && angular.isFunction(data)) {
-                        cb = data;
-                        data = null;
-                    }
-                    deferred.promise.then(cb);
-                    socket['legacy_' + methodName](url, data, function (result) {
-                        resolveOrReject(deferred, result);
-                    });
-                    return deferred.promise;
-                };
-            },
-            wrapEvent = function (eventName) {
-                socket['legacy_' + eventName] = socket[eventName];
-                socket[eventName] = function (event, cb) {
-                    if (cb !== null && angular.isFunction(cb)) {
-                        socket['legacy_' + eventName](event, function (result) {
-                            angularify(cb, result);
-                        });
-                    }
-                };
-            };
-
-        angular.forEach(httpVerbs, promisify);
-        angular.forEach(eventNames, wrapEvent);
-
-        return socket;
+    this.base = '';
+    this.config = {};
+    this.debug = false;
+
+    // like https://docs.angularjs.org/api/ng/service/$http#interceptors
+    // but with sails.io arguments
+    var interceptorFactories = this.interceptors = [
+      /*function($injectables) {
+         return {
+           request: function(config) {},
+           response: function(response) {},
+           requestError: function(config) {},
+           responseError: function(response) {}
+         };
+       }*/
+    ];
+
+    this.$get = ['$q', '$injector', '$rootScope', '$log', function($q, $injector, $rootScope, $log) {
+      var socket = (io.sails || io).connect(provider.url, provider.config);
+
+      // build interceptor chain
+      var reversedInterceptors = [];
+      angular.forEach(interceptorFactories, function(interceptorFactory) {
+        reversedInterceptors.unshift(
+          angular.isString(interceptorFactory) ?
+            $injector.get(interceptorFactory) : $injector.invoke(interceptorFactory));
+      });
+
+      // Send the request using the socket
+      function serverRequest(config) {
+        var defer = $q.defer();
+        if(provider.debug) $log.info('$sails ' + config.method.toUpperCase() + ' ' + config.url, config.data || '');
+
+        socket['legacy_' + config.method](config.url, config.data, function(result, jwr) {
+          // resolve promise if JSON web response is an object that has a statusCode 2xx
+          jwr.data = result; // backward compat, jwr.body also holds your data
```

Comment:
Looks like it will throw a ReferenceError around jwr being undefined with older versions of sails.io.  Maybe `jwr = jwr || {}` before this line